### DISCUSSION TITLE
Initial style guide content

### DISCUSSION THREAD

**Discussion by ngerakines** [lexicon-community/lexicon#7] [General]
This content will be formatted and used as a `LEXICON_STYLE_GUIDE.md` document.

### Data

When referencing lexicons in discussions, issues, and pull-requests consider the following:

* Run all JSON through a formatting tool for consistency

### Lexicon Style Guide

Reuse existing `community.lexicon.*` and `com.atproto.*` resources whenever possible.

Return the full record in "view" API endpoints.

Prefer open unions over closed unions. 

Leave space for arrays of labels in API responses.

Consider using empty response bodies to allow for future response body properties and attributes.

Avoid using period (".") characters in field names.

Consider the impact field names and formatting has on code generation tools across languages.

For consistency, consider using camelCase for field names.

The "defs" Lexicon file has emerged as a pattern that is used multiple types and functions within a Lexicon package level. I.e. `community.lexicon.calendar.defs`

Prefer single word nouns for record types and verb-prefixed strings for methods. `community.lexicon.calendar.event` and `community.lexicon.calendar.getEvents`

The ratio of byte strings to graphemes is roughly 10x. Use graphemes as real limits for real text and names; bytes-only good for identifiers (eg, DOI or ISBN)

The recommendation is that floats are stored as strings.

Using a format modifier for string types whenever possible.

Prefer "Strong Ref" types when linking to other types vs strings with an aturi format.

Leave room for additional properties and metadata when referencing blobs.

Leave room for multiple views (CDN links, representations, etc) of media files (blobs).

When an function returns a "view" along with a record, prefer "Strong Ref" types for record references.

Group additional data and metadata into typed objects. This allows additional and extra data to be more easily identified by the consumer of the function.

Prefer arrays of objects over arrays of strings

**Comment by tom-sherman** [comment]
Enshrining sidecar records as a recommended pattern for extension would be a good.

I wonder if there are use cases for sidecar records within the community.lexicon nsid 🤔

**Reply by ngerakines** [reply]
Can you define "sidecar record"? I'm pretty sure that I know what you mean, but I'd like to take what you give for it and put it in a terminology document of some sort.
